adapters/sspBC/sspbc.go

```diff
@@ -0,0 +1,339 @@
+	for _, imp := range imps {
+		if imp.ID == impID {
+			return imp.TagID
+		}
+	}
+
+	return impID
+}
+
+func (a *adapter) createBannerAd(bid openrtb2.Bid, ext responseExt, request *openrtb2.BidRequest, seat string) (string, error) {
+	var mcad mcAd
+
+	if strings.Contains(bid.AdM, "<!--preformatted-->") {
+		// Banner ad is already formatted
+		return bid.AdM, nil
+	}
+
+	// create McAd payload
+	mcad.Id = request.ID
+	mcad.Seat = seat
+	mcad.SeatBid = make([]openrtb2.SeatBid, 1)
+	mcad.SeatBid[0].Bid = make([]openrtb2.Bid, 1)
+	mcad.SeatBid[0].Bid[0] = bid
+
+	mcMarshalled, err := json.Marshal(mcad)
+	if err != nil {
+		return "", err
+	}
+
+	bannerData := &templatePayload{
+		SiteId:  ext.SiteId,
+		SlotId:  ext.SlotId,
+		AdLabel: ext.AdLabel,
+		PubId:   ext.PublisherId,
+		Page:    request.Site.Page,
+		Referer: request.Site.Ref,
+		McAd:    base64.URLEncoding.EncodeToString(mcMarshalled),
+	}
+
+	// Prepare banner html, using template file
+	var filledTemplate bytes.Buffer
+	if err := a.bannerTemplate.Execute(&filledTemplate, bannerData); err != nil {
+		return "", err
+	}
+
+	return filledTemplate.String(), nil
+}
+
+func getImpSize(imp openrtb2.Imp) string {
+	if imp.Banner == nil || len(imp.Banner.Format) == 0 {
+		return impFallbackSize
+	}
+
+	var (
+		areaMax int64
+		impSize = impFallbackSize
+	)
+
+	for _, size := range imp.Banner.Format {
+		area := size.W * size.H
+		if area > areaMax {
+			areaMax = area
+			impSize = fmt.Sprintf("%dx%d", size.W, size.H)
+		}
+	}
+
+	return impSize
+}
+
+func formatSspBcRequest(a *adapter, request *openrtb2.BidRequest) (*openrtb2.BidRequest, error) {
+	var siteID string
+	var isTest int
+
+	for i, imp := range request.Imp {
+		// read ext data for the impression
+		var extSSP openrtb_ext.ExtImpSspbc
+		var extBidder adapters.ExtImpBidder
+
+		// Read additional data for this imp.
+		// Errors here do not break the flow for this imp, and are ignored
+		if err := json.Unmarshal(imp.Ext, &extBidder); err == nil {
+			_ = json.Unmarshal(extBidder.Bidder, &extSSP)
+		}
+
+		// store SiteID
+		if extSSP.SiteId != "" {
+			siteID = extSSP.SiteId
+		}
+
+		// store test info
+		if extSSP.IsTest != 0 {
+			isTest = 1
+		}
+
+		// save current imp.id (adUnit name) as imp.tagid
+		// we will recover it in makeBids
+		imp.TagID = imp.ID
+
+		// if there is a placement id, use it in imp.id
+		if extSSP.Id != "" {
+			imp.ID = extSSP.Id
+		}
+
+		// check imp size and update e.ext - send pbslot, pbsize
+		// inability to set bid.ext will cause request to be invalid
+		impSize := getImpSize(imp)
+		impExt := requestImpExt{
+			Data: adSlotData{
+				PbSlot: imp.TagID,
+				PbSize: impSize,
+			},
+		}
+
+		impExtJSON, err := json.Marshal(impExt)
+		if err != nil {
+			return nil, err
+		}
+		imp.Ext = impExtJSON
+		// save updated imp
+		request.Imp[i] = imp
+	}
+
+	// update site info (ID, of present, and request domain)
+	if siteID != "" {
+		request.Site.ID = siteID
```

Comment:
`Site` is shared memory and cannot be directly modified. You need to make a copy of it first and then modify, such as:

```
siteCopy := *request.Site
request.Site = &siteCopy
request.Site.ID = siteID
``` 

This would be detected by our json tests if you add a test file where the `site.id` is empty or different.